src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -62,11 +73,105 @@ private static IEnumerable<string> GetTagNames(Type tagHelperType)
                     name = name.Substring(0, name.Length - TagHelperNameEnding.Length);
                 }
 
-                return new[] { ToHtmlCase(name) };
+                return new[]
+                {
+                    new ElementTarget
+                    {
+                        TagName = ToHtmlCase(name),
+                        AttributeNames = Enumerable.Empty<string>()
+                    }
+                };
             }
 
-            // Remove duplicate tag names.
-            return attributes.SelectMany(attribute => attribute.Tags).Distinct();
+            return targetElementAttributes.SelectMany(
+                targetElementAttribute =>
+                {
+                    IEnumerable<string> tagNames;
+                    IEnumerable<string> attributeNames = null;
+
+                    if (!TryGetValidatedNames(
+                            targetElementAttribute.Tags,
+                            pascalNameTarget: "Tag",
```

Comment:
1. "tag" and "attribute" should be localized
2. `pascalNameTarget` is horrible
so, create separate resources for attribute and tag validation issues.  then pass in a `bool targettingAttributes` parameter to the validation method.